### DISCUSSION TITLE
Authorized GraphQL request in loader?

### DISCUSSION THREAD

**Discussion by shkim** [remix-run/remix#4806] [Q&A]
I want to use the result data from GraphQL request (to different server) in loader.
Currently, I'm using URQL but I think this problem is also same with Apollo Client.

Before using Urql query/mutation in any route component,
Urql client should be initialized first and then it must be provided by <Provider> component in upper level.
(Very common pattern in most React library)

1. The problem is React context is null in loader, so useQuery or any use** are not available.
2. At the moment, I cannot use a global GraphQL client instance because I need an authorized client, it must not be shared among different browsers.

I could not find any example which shows how to use some popular(full-fledged) GraphQL client in Remix.
Can I get a solution for this situation?

**Comment by shkim** [comment]
According to this Urql guide: https://formidable.com/open-source/urql/docs/advanced/server-side-rendering/#ssr-with-getstaticprops-or-getserversideprops
I think I should create a new GraphQL client (with auth cookie) in every loaders.
(For server side, it seems that gql client instance cannot be shared through components rendering.)

And another approach is to pass the initial states to client from prerender result in server
(Example: https://www.apollographql.com/blog/apollo-client/how-to-use-apollo-client-with-remix )
I managed to do last project this way with Vue 3 + Vite SSR + Urql, but that experience is why I'm seeking another tech stack like Remix. -- I don't wanna try this again with Remix.

Anyway, at the moment, I concluded that I cannot get the server-side GraphQL client instance which is initialized before in entry.server.tsx or root.tsx.

**Reply by dusty** [reply]
You can use fetch in your loader to do the graphql query and pass the response to the client via useLoaderData.  

When we built a new app in remix, we found that urql, apollo, etc. just wasn't needed anymore for us.  Remix will revalidate your loaders after an action, which will refresh your data.  Or if you want to change the query, you could just use search params in the url to re-run the loader.

Perhaps this will work for you as well.

Here is an example using gql, although graphql-codegen would also be a nice solution to use for generating the queries.  This is using a Bearer token for auth.


```
// some-route.tsx
import gql from "graphql-tag";
import fetchGql from '~/lib'

async function loader({ request }) {
  const query = gql`
    query user(id: String) {
      user(id: $id) {
        firstName
        lastName
      }
    }
  `;

  return fetchGql({
    operationName: "user",
    query,
    variables: { id: 5 },
    request,
  });
}

export default function MyView() {
  const { data, errors } = useLoaderData();
  //...
}
```

```
// lib/fetch-gql.server.ts
import { print } from "graphql";
import { getUserSession } from '~/lib'

export async function fetchGql({ operationName, query, variables, request }) {
  const session = await getUserSession(request);
  const token = session.get("token");

  const headers = new Headers();
  headers.set("content-type", "application/json");
  if (token) headers.set("authorization", `Bearer ${token}`);

  const config: RequestInit = {
    headers,
    method: "POST",
    body: JSON.stringify({
      operationName: "user",
      query: print(query),
      variables: { id: 5 },
    }),
  };

  const res = await fetch(process.env.GRAPHQL_URL, config);
  const data = await res.json();

  // network errors can go to a catch boundary
  if (!res.ok) {
    throw new Response(res.statusText, {status: res.status || 500});
  }

  return data;
}
```

**Comment by KevinMind** [comment]
I wonder if we could have first class support for graphql loader/action with a simple fetch wrapper? Shouldn't this be trivial to implement. 

Remix provides 80% of what graphql clients do and with some conventions on mapping search params and formData elements to gql variables this should be 100% possible without any extra dependencies.

**Reply by KevinMind** [reply]
Exactly like @dusty is suggesting. I would love to see a codegen plugin for remix. I might give this a shot myself.

**Reply by dusty** [reply]
I ended up using graphql-codegen using the typed document node plugin.  It's quite nice and easy to use because you use a typed document instead of many different generated hooks.  I'm out of town this week, but I'll drop in an example when I get back.

It looks something like this when I use it.  

```
import { GetThingsDocument } from '~/graphql'
import { fetchGql } from '~/lib'

export async function loader({ request }: LoaderArgs) {
  const query = await fetchGql(GetThingsDocument, { some: vars }, request)
  return json({things: query.getThings })
}

export default function ThingsRoute() {
  const things = useLoaderData<typeof loader>()
  ...
}
```
